---
title: "B. germanica microbiome analysis in R"
author: "Zachery Wolfe", "Kim McFarland"
date: "2021"
output:
  html_document: 
    toc: true
    toc_float: true
  pdf_document: default
editor_options: 
  chunk_output_type: console
always_allow_html: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

knitr::opts_knit$set(root.dir = '~/C:/Users/xcwol/Downloads/ANSC595-project-German-cockroach-gut-microbiome-master/ANSC595-project-German-cockroach-gut-microbiome-master/microbiome files"')
```

```{r}
source("https://bioconductor.org/biocLite.R")
BiocManager::install()
BiocManager("phyloseq")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.12")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("phyloseq")
```

# Load packages
```{r Load packages, results="hide"}
library(ape)
library(dplyr)
library(ggplot2)
library(gplots)
library(lme4)
library(phangorn)
library(phyloseq)
library(plotly)
library(tidyr)
library(vegan)
library(VennDiagram)
```

##Load Data
##See "Microbiota_Analysis_in_R_setup.R"

##Clean up the data

###OTU table
We need to use the "Group" column as the row names so that it will match our metadata
```{r}
row.names(OTU) = OTU$Group
```

We then need to remove the "label", "numOTUs", and "Group" columns as they are not OTU counts
```{r}
OTU.clean = OTU[,-which(names(OTU) %in% c("label", "numOtus", "Group"))]
```

###Taxonomy table
For the taxonomy table, we name the rows by the OTU #
```{r}
row.names(tax) = tax$OTU
```

Remove all the OTUs that don't occur in our OTU.clean data set
```{r}
tax.clean = tax[row.names(tax) %in% colnames(OTU.clean),]
```

We then need to separate the "taxonomy" column so that each level (*i.e.* Domain, Phylum, etc) is in it's own column.
```{r}
tax.clean$Taxonomy <- gsub("[(]\\d*[)]",replacement="",x=tax.clean$Taxonomy)
tax.clean = separate(tax.clean, Taxonomy, into = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain"), sep=";")
```

Finally, we remove the "Size" and "Strain" columns as well as "OTU" since these are now the row names
```{r}
tax.clean = tax.clean[,-which(names(tax.clean) %in% c("Size", "Strain", "OTU"))]
tax.clean2 <- gsub("[(]\\d*[)]",replacement="",x=tax.clean)
```

###Metadata and SCFA tables
These tables do not require any modification since I created them in Excel exactly as I need them for this R analysis. 
##Order the data
To make viewing and using the data easier, we will make sure our tables have samples (rows) in the same order. Since OTU.clean, meta, and SCFA have sample names as row names, we order by these.
```{r Order the data}
OTU.clean = OTU.clean[order(row.names(OTU.clean)),]
```

##Set seed

```{r}
set.seed(34343)
```

#Alpha-diversity

##Explore alpha metrics
You want the data to be roughly normal so that you can run ANOVA or t-tests. If it is not normally distributed, you will need to consider non-parametric tests such as Kruskal-Wallis.

Here, we see that none of the data are normally distributed. This occurs with the subset but not the full data set because I've specifically selected samples with divergent alpha metrics. In general, you will see roughly normal data for Shannon's diversity as well as most richness metrics. Simpson's diversity, on the other hand, is usually skewed as seen here.
```{r}
#Create 2x2 plot environment 
par(mfrow = c(2, 2))

#Plots
hist(metainvsimpson$invsimpson, main="Inverse Simpson diversity", xlab="", breaks=10)
hist(metashannon$shannon, main="Shannon diversity", xlab="", breaks=10)
```

Test for normalcy statistically
```{r}
shapiro.test(metainvsimpson$invsimpson)
shapiro.test(metashannon$shannon)
```

We see that, as expected from the graphs, none are normal.

##Categorical variables

**Normally distributed metrics**

Does treatment or strain impact the Shannon diversity of the fecal microbiota?
```{r}
#Run the ANOVA and save it as an object
aov.shannon.interaction = aov(shannon ~ Treatment*Strain, data=metashannon)
aov.shannon.traits = aov(shannon ~ Treatment+Strain, data=metashannon)
aov.invsimpson.interaction = aov(invsimpson ~ Treatment*Strain, data=metainvsimpson)
aov.invsimpson.traits = aov(invsimpson ~ Treatment+Strain, data=metainvsimpson)
#Call for the summary of that ANOVA, which will include P-values
summary(aov.shannon.interaction)
summary(aov.shannon.traits)
summary(aov.invsimpson.interaction)
summary(aov.invsimpson.traits)
```

Run Tukey's honest significance test of ANOVA.
```{r}
TukeyHSD(aov.shannon.interaction)
TukeyHSD(aov.invsimpson.interaction)
```

Plot the data
```{r}
#Return the plot area to 1x1
par(mfrow = c(1, 1))
#Plot
boxplot(shannon ~ Interaction, data=metashannon, ylab="Shannon's diversity")
boxplot(invsimpson ~ Interaction, data=metainvsimpson, ylab="Inverse Simpson's diversity")

shannon <- ggplot(metashannon, aes(Interaction, shannon)) + 
  geom_boxplot(aes(color = Interaction)) + 
  ylim(c(0,6)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
plot(shannon)

invsimpson <- ggplot(metainvsimpson, aes(Interaction, invsimpson)) + 
  geom_boxplot(aes(color = Interaction)) + 
  ylim(c(0,50)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
plot(invsimpson)

```

**Non-normally distributed metrics**

Treatment is categorical; we use Kruskal-Wallis (non-parametric equivalent of ANOVA). If we have only two levels, we would run Wilcoxon rank sum test (non-parametric equivalent of t-test)
```{r}
kruskal.test(invsimpson ~ Interaction, data=metainvsimpson)
kruskal.test(shannon ~ Interaction, data=metashannon)
```

We can test pairwise within the age groups with Wilcoxon Rank Sum Tests. This test has a slightly different syntax than our other tests
```{r}
pairwise.wilcox.test(metainvsimpson$invsimpson, metainvsimpson$Interaction, p.adjust.method="fdr")
```

#Beta-diversity

####Dot plots
Calculate the nMDS values for a 2-axis `k=2` graph using the OTU-based Bray-Curtis metric that takes into account both the presence/absence and abundance of OTUs in your samples. This uses the `metaMDS` function from the package `vegan`.
```{r}
BC.nmds = metaMDS(OTU.clean, distance="bray", k=2, trymax=1000)
Jac.nmds = metaMDS(OTU.clean, distance="jaccard", k=2, trymax=1000)
```

Then plot the nMDS with different colors for your different groups of interest, add colors
```{r}
#if running line-by-line run this entire block at one time.

par(mfrow = c(1, 1))
#Create a blank plot for the nmds
plot(BC.nmds$points, type="p", main="Bray-Curtis",xlim = c(-1.0, 1.0), ylim = c(-1.5, 1.5)) 
#Add the points colored by group
points(BC.nmds, display="sites", pch=20, col=c("Red", "Red", "Red","Magenta","Magenta","Magenta","Blue","Blue","Blue","cyan","cyan","cyan"))
#Add a legend
legend(-0.75, 0.45, legend=c("Dan Ctrl","Dan Kan","Jw Ctrl","Jw Kan"), col=c("Red","Magenta","Blue","cyan"), pch=20)

plot(Jac.nmds$points, type="p", main="Jaccard",xlim = c(-1.0, 1.0), ylim = c(-1.5, 1.5))
#Add the points colored by group
points(Jac.nmds, pch=20, col=c("Red", "Red", "Red","Magenta","Magenta","Magenta","Blue","Blue","Blue","cyan","cyan","cyan"))
#Add a legend
legend(0.75, 0.45, legend=c("Dan Ctrl","Dan Kan","Jw Ctrl","Jw Kan"), col=c("Red","Magenta","Blue","cyan"), pch=20)
```

####3D plots
If your stress is high (like over 0.3) for your `metaMDS` calculation, you probably need to increase to 3 axes `k=3`. Graphing a 3D plot is much more complicated, and there are a number of packages that could be used. Here, we will use one option from the `plotly` package to visualize a 3D Bray-Curtis plot.

```{r}
#Calculate the Bray-Curtis nMDS for 3-axis
BC.nmds.3D = metaMDS(OTU.clean, distance="bray", k=3, trymax=1000)
```

Extract x-y-z values for this nmds
```{r}
BCxyz = scores(BC.nmds.3D, display="sites")
BCxyz
```

Plot the xyz coordinates and color by group
```{r}
plot_ly(x=BCxyz[,1], y=BCxyz[,2], z=BCxyz[,3], type="scatter3d", mode="markers", color=metashannon$group[1:12], colors=c("Red","magenta","Blue","cyan"))
```

###Phylogentic-based metrics
The most common of this type of beta-diversity metrics is UniFrac.

####Create physeq object
To start, you must make a `phyloseq` object which includes the OTU.clean, meta, and tax.clean data. We tell R which tables are each type
```{r nMDS_unifrac}
OTU.UF = otu_table(as.matrix(OTU.clean), taxa_are_rows=FALSE)
tax.UF = tax_table(as.matrix(tax.clean))
metashannon.UF = sample_data(metashannon)
metainvsimpson.UF = sample_data(metainvsimpson)
#Run chunk below before running @sample_data
physeqshannon@sam_data = sample_data(metashannon)
```

We then merge these into an object of class phyloseq.
```{r}
physeqtest = phyloseq(tax.UF, OTU.UF)
physeqshannon = phyloseq(OTU.UF, tax.UF,sample_names(metashannon.UF))
physeqinvsimpson = phyloseq(OTU.UF, tax.UF,sample_names(metainvsimpson.UF))

physeqshannonb <- subset_taxa(physeqshannon, Family != "Blattabacteriaceae")
physeqshannonb <- prune_samples(sample_sums(physeqshannonb)>=1, physeqshannonb)

physeqinvsimpsonb <- subset_taxa(physeqshannon, Family != "Blattabacteriaceae")
physeqinvsimpsonb <- prune_samples(sample_sums(physeqshannonb)>=1, physeqshannonb)
```

To add the phylogenetic component to UniFrac, we calculate a rooted phylogenetic tree of our OTUs. This takes a long time so we have provided the tree for you (See NJ.Tree.R).

Kim Dill-McFarland has pre-calculated this tree at "https://github.com/kdillmcfarland/workshops_UW_Madison/blob/master/Microbiota_analysis_R/Data/NJ.tree.RData"

You can load it after downloading it with
```{r}
load("NJ.tree.Rdata")
```

Add this tree to your physeq object. This object will be what is used in UniFrac calculations.
```{r}
physeqshannon.tree = merge_phyloseq(taxa_names(physeqshannonb), NJ.tree)
physeqinvsimpson.tree = merge_phyloseq(taxa_names(physeqinvsimpsonb), NJ.tree)
```

We can look at this object and see its components.
```{r}
physeqshannon.tree
physeqinvsimpson.tree
```

```{r}
plot_tree(physeqshannon.tree, label.tips="taxa_names", ladderize="left", plot.margin=0.3)
plot_heatmap(physeqshannonb, taxa.label = "Genus")
## top20OTU is created in line 496. Either run the chunk on line 496 first or change the heatmap plot below
plot_heatmap(top20OTU, taxa.label = "Genus",low="white", high="purple", na.value="grey")


plot_bar(physeqshannonb, x="Sample", fill="Phylum") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack")
plot_bar(physeqshannonb, x="Sample", fill="Family") + geom_bar(aes(color=Family, fill=Family), stat="identity", position="stack")

plot_bar(physeqshannonb, x="Sample",y="Abundance", fill = "Family")
```

```{r}
# Set colors for plotting
my_colors <- c(
  '#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c',
  '#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928', 
  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
   "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
  "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861", "black"
)

#If you want different taxonomic level, find and replace the taxonomic level listed here
my_level <- c("Genus")
#rm(taxa.summary)

ml ="Genus"
for(ml in my_level){
  print(ml)

  
  taxa.summary <- physeqinvsimpsonb %>%
    tax_glom(taxrank = ml, NArm = FALSE) %>%  # agglomerate at `ml` level
    transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
    psmelt()  %>%                               # Melt to long format
    group_by(Sample, get(ml)) %>%
    summarise(Abundance.average=mean(Abundance),)
    names(taxa.summary)[2] <- ml
  
  physeq.taxa.average <- taxa.summary %>% 
    group_by(get(ml)) %>%
    summarise(overall.average=mean(Abundance.average))
  names(physeq.taxa.average)[1] <- ml
  
  # merging the phyla means with the metadata #
  physeq_meta <- merge(taxa.summary, physeq.taxa.average)
    
  abund_filter <- 0.01
  physeq_meta_filtered <- filter(physeq_meta, overall.average>abund_filter)
  str(physeq_meta_filtered)
  
  physeq_meta_filtered$group = factor(physeq_meta_filtered$Sample, c("Dan-Ctrl", "Dan-Kan","Jwax-Ctrl","Jwax-Kan"))
  
  # Plot 
  ggplot(physeq_meta_filtered, aes(x = Sample, y = Abundance.average, fill = get(ml))) + 
    #facet_grid(.~subject) +
    geom_bar(position="fill",stat = "identity") +
    scale_fill_manual(values = my_colors) +
    # Remove x axis title
    theme(axis.title.x = element_blank()) + 
    ylim(c(0,1)) +
    guides(fill = guide_legend(reverse = F, keywidth = .5, keyheight = .5, ncol = 1)) +
    theme(legend.text=element_text(size=8)) +
    #theme(legend.position="bottom") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    theme(legend.title = element_blank()) +
    ylab("Relative Abundance") +
    ggtitle(paste0(ml, " (>", abund_filter * 100,"%) Composition of microbiome samples")) 
}
```

##Statistically test beta-diversity
While nMDS gives us a visual of beta-diversity, it does not test for statistical differences. We do this with permutational analysis of variance (PERMANOVA). This tests whether the overall microbial community differs by a variable of interest. You can run them with Bray-Curtis, Jaccard, weighted or unweighted UniFrac.

###PERMANOVA

Calculate distance and save as a matrix. **Note**: adonis cannot handle NA or blanks in data. Subset to only samples with complete metadata before running `vegdist`.
```{r}
BC.dist=vegdist(OTU.clean, distance="bray")
#Run PERMANOVA on distances.
adonis(BC.dist ~ Treatment*Strain, data = metashannon, permutations = 1000)
adonis(BC.dist ~ Treatment*Strain, data = metainvsimpson, permutations = 1000)
```

Similarly for Jaccard
```{r}
J.dist=vegdist(OTU.clean, distance="jaccard")
adonis(J.dist ~ Treatment*Strain, data = metashannon, permutations = 1000)
adonis(J.dist ~ Treatment*Strain, data = metainvsimpson, permutations = 1000)
```

##Beta dispersion

Calculate dispersion (variances) within each group.
```{r}
disp.age = betadisper(BC.dist, metashannon$Interaction)
```

Perform an ANOVA-like test to determine if the variances differ by groups.
```{r}
permutest(disp.age, pairwise=TRUE, permutations=1000)
```

Combine this with plot
```{r}
plot(BC.nmds, type="n", main="Bray-Curtis")
legend(.6,-2, legend=c("Danville_Acetone","Danville_Kanamycin","J-wax_Acetone","J-wax_Kanamycin"), col=c("red","magenta","blue","cyan"), pch=20)
ordiellipse(BC.nmds, groups=metashannon$Interaction, display="sites", kind="se", conf=0.99, label=FALSE, col="red", draw="polygon", alpha=200, show.groups = c("Danville_Acetone"), border=FALSE)
ordiellipse(BC.nmds, groups=metashannon$Interaction, display="sites", kind="se", conf=0.99, label=FALSE, col="magenta", draw="polygon", alpha=200, show.groups = c("Danville_Kanamycin"), border=FALSE)
ordiellipse(BC.nmds, groups=metashannon$Interaction, display="sites", kind="se", conf=0.99, label=FALSE, col="blue", draw="polygon", alpha=200, show.groups = c("J-wax_Acetone"), border=FALSE)
ordiellipse(BC.nmds, groups=metashannon$Interaction, display="sites", kind="se", conf=0.99, label=FALSE, col="cyan", draw="polygon", alpha=200, show.groups = c("J-wax_Kanamycin"), border=FALSE)
```

#Other visualizations
##Bar charts

Let's explore some bar chart options. First, we'll make the classic additive bar chart for phyla in our samples
```{r Bar_charts}
plot_bar(physeqshannonb, fill="Genus")
```

Remove the lines between OTUs in the bars
```{r}
plot_bar(physeqshannonb, fill="Genus") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack")
```

And only showing the top xxx most abundant phyla
```{r}
#Sort the Phyla by abundance and pick the top 5
top5P.names = sort(tapply(taxa_sums(physeqinvsimpsonb), tax_table(physeqinvsimpsonb)[, "Phylum"], sum), TRUE)[1:5]
top10P.names = sort(tapply(taxa_sums(physeqinvsimpsonb), tax_table(physeqinvsimpsonb)[, "Phylum"], sum), TRUE)[1:10]
top5G.names = sort(tapply(taxa_sums(physeqinvsimpsonb), tax_table(physeqinvsimpsonb)[, "Genus"], sum), TRUE)[1:5]
top10G.names = sort(tapply(taxa_sums(physeqinvsimpsonb), tax_table(physeqinvsimpsonb)[, "Genus"], sum), TRUE)[1:10]
#Cut down the physeq.tree data to only the top 10 Phyla
top5P = subset_taxa(physeqinvsimpsonb, Phylum %in% names(top5P.names))
top5G = subset_taxa(physeqinvsimpsonb, Genus %in% names(top5G.names))
top10P = subset_taxa(physeqinvsimpsonb, Phylum %in% names(top10P.names))
top10G = subset_taxa(physeqinvsimpsonb, Genus %in% names(top10G.names))
#Plot
plot_bar(top10P, fill="Phylum") + geom_bar(aes(fill=Phylum), stat="identity", position="stack")
plot_bar(top10G, fill="Genus") + geom_bar(aes(fill=Genus), stat="identity", position="stack") + scale_color_brewer(palette="Rainbow")
```

Another way to simplify these bar plots is to not show all OTUs for one sample in one bar. We can do this with facet_grid
```{r}
plot_bar(top10P, fill="Phylum", facet_grid = ~Phylum) + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack")

plot_bar(top5G, fill="Genus", facet_grid = ~Genus) + geom_bar(aes(color=Genus, fill=Genus), stat="identity", position="stack")

plot_bar(top5G, x="Genus", fill="Genus", facet_grid = ~Genus) + geom_bar(aes(fill=Genus), stat="identity", position="stack")
```

##Heat maps

###OTUs
We're going to just look at the 20 most abundant OTUs to make it more readable.
```{r Heat_maps}
#Sort OTUs by abundance and pick the top 20
top20OTU.names = names(sort(taxa_sums(physeqinvsimpsonb), TRUE)[1:20])
#Cut down the physeq.tree data to only the top 10 Phyla
top20OTU = prune_taxa(top20OTU.names, physeqinvsimpsonb)

top20OTU
top20OTU@sam_data = sample_data(metashannon)
```

First, make a heatmap of OTU abundance across all samples
```{r}
plot_heatmap(top20OTU)
```

```{r}
plot_heatmap(top20OTU, sample.label="group", sample.order="Interaction")
```

We can label the OTU taxa
```{r}
plot_heatmap(top20OTU, sample.label="group", sample.order="Interaction", taxa.label="Genus")
```

And group OTUs within the same Phyla
```{r}
plot_heatmap(top20OTU, sample.label="group", sample.order="Interaction", taxa.label="Genus", taxa.order="Phylum")
```

We can also change the colors (white -> purple), including the 0s/NAs (grey).
```{r}
plot_heatmap(top20OTU, sample.label="group", sample.order="Interaction", taxa.label="Genus", taxa.order="Phylum", low="white", high="red", na.value="grey")
```

```{r}
plot_heatmap(top20OTU, "NMDS", "bray", title="Bray-Curtis")
```

###Beta-diversity

We do not want to use the plot_heatmap() function from `phyloseq` because it requires the input of a physeq object. Instead, we can use our distance matrices as inputs for a `gplots` command. This command will automatically group samples by similarity (trees)
```{r}
#Bray-Curtis
heatmap.2(as.matrix(BC.dist))
```

You could also change colors 
```{r}
#Rainbow colors
rc <- rainbow(nrow(as.matrix(BC.dist)), start=0, end=0.9)
heatmap.2(as.matrix(BC.dist), col=rc)
```

###RANDOM FOREST analysis
from Michelle Berry
https://rpubs.com/michberr/randomforestmicrobe

```{r}
install.packages("randomForest")
library(randomForest)
#Check how many OTUs you have. If you have more than 1000, you may want to filter out rare OTUs, as shown below.
ntaxa(physeqshannonb)
ntaxa(physeqinvsimpsonb)
# Set prunescale 
prunescale = 0.0001
seq_depth = sample_sums(physeqshannonb)[1]
# Prune out rare OTUs by mean relative abundance set by prunescale
tax.mean <- taxa_sums(physeqshannonb)/nsamples(physeqshannonb)

#Select the samples we want to compare

#Thor, How to run this and how to get the top 300 taxa??? show heatmap blogpost

rf.samples = c("Control", "Kanamycin")
physeq.prune <- prune_taxa(tax.mean > prunescale*seq_depth, physeqshannonb)
physeq.prune <- prune_taxa(tax.mean > prunescale*seq_depth, physeqshannonb)
physeq.prune@otu_table = physeq.prune@otu_table * 2 + 1
physeq.prune

# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- (otu_table(physeq.prune))
dim(predictors)
```

```{r}
# Make one column for our outcome/response variable 
response <- as.factor(sample_data(physeq.prune)$Treatment)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)
```

Now we will use the randomForest package to train and test our random forest model using the “out of bag” error to estimate our model error.

Results
```{r}
set.seed(23456)

physeq.classify <- randomForest(response~., data = rf.data, ntree = 100)
print(physeq.classify)
```

In most statistical learning algorithms, the data needs to be split up into “training” and “test” data. The idea is to train the model on one set of data and test it on a naive set of data. Random forests are nice because you have a built-in way of estimating the model error. Since only ~2/3 of the data is used everytime we bootstrap our samples for construction of the kth tree, we can use the remaining ~1/3 of the data (called the out of bag samples) to test model error: Our out of bag error is 6.25% .

# What variables are stored in the output?
```{r}
names(physeq.classify)
```

###plots
Lets make some plots of the most important variables in our model. For a classification tree, variable importance is measured by mean decrease in GINI coefficient (measure of node purity).

# Make a data frame with predictor names and their importance
```{r}
imp <- importance(physeq.classify)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort <- arrange(imp, desc(MeanDecreaseGini))
imp.sort$predictors <- factor(imp.sort$predictors, levels = imp.sort$predictors)

# Select the top 10 predictors
imp.20 <- imp.sort[1:20, ]

# ggplot
ggplot(imp.20, aes(x = predictors, y = MeanDecreaseGini)) +
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle(paste("Most important OTUs for classifying between", rf.samples[1], "and ", rf.samples[2]))

# What are those OTUs?
otunames <- imp.20$predictors
r <- rownames(tax_table(physeqshannon)) %in% otunames
tax_table(physeqshannon)[r, ]
```

###Differential Abundance with DESeq2

Adapted from https://joey711.github.io/phyloseq-extensions/DESeq2.html

First load DESeq2.
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))

    install.packages("BiocManager")

 
BiocManager::install("DESeq2")
```

To use DESeq, we need no zeros in our OTU table. So we will edit the table by multiplying by 2 and + 1
```{r}
OTU.clean2 <- OTU.clean * 2 + 1
```

Now make the phyloseq object:
```{r}
OTU.physeq = otu_table(as.matrix(OTU.clean2), taxa_are_rows=FALSE)
tax.physeq = tax_table(as.matrix(tax.clean))
meta.physeq = sample_names(metashannon)
```

We then merge these into an object of class phyloseq.
```{r}
physeq_deseq = phyloseq(OTU.physeq, tax.physeq, meta.physeq)
```

The function phyloseq_to_deseq2 converts phyloseq-format microbiome data into a DESeqDataSet with dispersions estimated using the experimental design formula.
```{r}
library("DESeq2")

physeq_deseq@sam_data = sample_data(metashannon)
diagdds = phyloseq_to_deseq2(physeq_deseq, ~ Treatment)
diagdds = DESeq(diagdds, test="Wald", fitType="parametric")
#the test type of "Wald" tests for significance of coefficients in a Negative Binomial GLM
```

###Investigate test results table

```{r}
my_contrast = c("Treatment","Control", "Kanamycin")
res = results(diagdds, contrast = my_contrast, cooksCutoff = FALSE)

alpha = 0.01
sigtab = res[which(res$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(physeq_deseq)[rownames(sigtab), ], "matrix"))
head(sigtab)
```

Volcano Plot
```{r}
#reset par
par(mfrow=c(1,1))
# Make a basic volcano plot
with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-15,15)))

# Add colored points: blue if padj<0.01, red if log2FC>1 and padj<0.05)
with(subset(res, padj<.01 ), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, padj<.01 & abs(log2FoldChange)>2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
```

The following makes a nice ggplot2 summary of the results.
```{r}
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))
DESeq_fig = ggplot(sigtab, aes(x=Genus, y = log2FoldChange, color=Phylum)) + geom_point(size=3) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
plot(DESeq_fig)
```

#########################################################################